Use constant time string comparisons

brendanlong opened the pull request:
This uses constant time string comparisons everywhere that looked plausibly like it
might matter:

 - Plaintext password comparison
 - Digest nonce comparison
 - Digest hash comparison

Fixes #82

brendanlong commented:
I'm not sure if I'm being overly careful about None checking..

miguelgrinberg commented:
Your change breaks the Python 2.7 build (there is no compare_digest before version 2.7.7 or on pypy). Take a look at how another project implemented this with a fallback: https://github.com/scaleway/python2-secrets/blob/develop/secrets/secrets.py#L32-L42.

brendanlong commented:
I pushed a new version that adds a fallback. I included a function comment to explain why it's written the way it is, but you can see more details here: https://github.com/brendanlong/python-constant-time-string-stats

brendanlong commented:
Looks like whatever we use needs to convert both strings to the same format since `==` was allowing us to compare bytes and strings before?

brendanlong commented:
Actually I messed something up in my comparisons. The for Python 2.6 are:

`==`: ~250 ns
Loop and xor each byte: ~10 μs
`md5(a) == md5(b) and a == b`: ~3 μs
Comparing salted HMACs: ~38 μs

If the performance doesn't matter very much, I'm pretty confident that the salted HMAC version doesn't leak anything. If performance does matter, the unsalted version is the fastest thing that works on Python 2.6 and only leaks the hash if the attacker is capable of an arbitrary length chosen-prefix attack.

miguelgrinberg commented:
@brendanlong do you plan to fix the failures in the latest version of your PR, or would you like me to look into it myself? I can't really merge it in this state, Python 2 appears broken.

brendanlong commented:
I just updated this with a version that works for me on Python 2.6, 2.7, 3.6, and 3.7.

The issue was bytes/unicode mismatches. I added some code to handle this in the password comparison case. With that change, everything works with string or bytes passwords, but this still requires hashes to be bytes. I'm not sure if I should also add some code to convert unicode hashes to bytes?

brendanlong commented:
The complicated part is actually `compare_digest` in Python 2.7, since we end up with `client_password` being unicode and `stored_password` being str/bytes. I switched to a simpler fallback though since I think it does make it simpler if we don't need to care about the types.

brendanlong commented on flask_httpauth.py:
```diff
@@ -142,8 +154,18 @@ def authenticate(self, auth, stored_password):
+        try:
+            client_password = client_password.encode("utf-8")
+        except AttributeError:
+            pass
```
We still need all of this stuff to ensure both passwords are the same type (the easiest way is to just convert both to bytes).

miguelgrinberg replied:
Why would they be of different types? If they are different types then the current version of the package would report the passwords as non matching. Or am I missing something?
